src/persist-client/src/impl/state.rs

```diff
@@ -273,13 +273,16 @@ where
         Ok((work_ret, new_state))
     }
 
-    pub fn snapshot(&self, as_of: &Antichain<T>) -> Result<Vec<String>, InvalidUsage> {
+    pub fn snapshot(
+        &self,
+        as_of: &Antichain<T>,
+    ) -> Result<Result<Vec<String>, Upper<T>>, Since<T>> {
         if PartialOrder::less_than(as_of, &self.collections.since) {
-            return Err(InvalidUsage(anyhow!(
-                "snapshot with as_of {:?} cannot be served by shard with since: {:?}",
-                as_of,
-                self.collections.since
-            )));
+            return Err(Since(self.collections.since.clone()));
+        }
+        let upper = self.collections.upper();
+        if PartialOrder::less_equal(&upper, as_of) {
+            return Ok(Err(Upper(upper)));
         }
```

Comment:
It seems like this will prevent any response if `since == upper` -- which as far as I can tell is the initial state for a new collection -- where both equal `Timestamp::minimum()`

I imagine it's more ergonomic to handle that case internally rather than each caller doing a check but what do you think?

Reply:
as far as I understand this behavior is correct (though I agree that it's quite unintuitive). `as_of` is inclusive, while `upper` is exclusive. concretely, this indeed means that a newly initialized shard with upper of `antichain[T::minimum()]` cannot be read as_of `antichain[T::minimum()]`. fwiw, none of this is coming from me. this is what is prescribed by the formalism

in the maelstrom PR, I "initialize" the shard when a process first starts by attempting a cas from `antichain[T::minimum()]` to `antichain[T::minimum()+1]`. this either sets it up the shard to an initial upper of `antichain[T::minimum()+1]` allowing reads as_of `antichain[T::minimum()]`, or if the shard has already been used, returns the upper frontier information. either way, we have an upper now. https://github.com/MaterializeInc/materialize/pull/12108/files#diff-bfde6fe4d24d0f4f78767d45c96169d046ebd064519ab2bc736719ba88825264R125-R144